controllers/vaultdynamicsecret_controller.go

```diff
@@ -441,6 +442,100 @@ func (r *VaultDynamicSecretReconciler) syncSecret(ctx context.Context, c vault.C
 	return secretLease, true, nil
 }
 
+// awaitVaultSecretRotation waits for the Vault secret to be rotated. This is
+// necessary for the case where the Vault secret is a static-creds secret and includes
+// a rotation schedule.
+func (r *VaultDynamicSecretReconciler) awaitVaultSecretRotation(ctx context.Context, o *secretsv1beta1.VaultDynamicSecret,
+	c vault.ClientBase, lastResponse vault.Response) (*secretsv1beta1.VaultStaticCredsMetaData,
+	vault.Response,
+	error,
+) {
+	logger := log.FromContext(ctx).WithName("awaitVaultSecretRotation")
+
+	resp := lastResponse
+	respData := lastResponse.Data()
+	staticCredsMeta, err := vaultStaticCredsMetaDataFromData(respData)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// if we are not handling static creds or the rotation schedule is not set, then
+	// we can return early.
+	if !r.isStaticCreds(staticCredsMeta) || staticCredsMeta.RotationSchedule == "" {
+		return staticCredsMeta, resp, nil
+	}
+
+	lastSyncStaticCredsMeta := o.Status.StaticCredsMetaData.DeepCopy()
+	inLastSyncRotation := lastSyncStaticCredsMeta.LastVaultRotation == staticCredsMeta.LastVaultRotation
+	switch {
+	case !inLastSyncRotation:
+		// return early, not in the last rotation
+		return staticCredsMeta, resp, nil
+	case lastSyncStaticCredsMeta.RotationSchedule == "":
+		// return early, rotation schedule was not set in the last sync
+		return staticCredsMeta, resp, nil
+	case lastSyncStaticCredsMeta.RotationSchedule != staticCredsMeta.RotationSchedule:
+		// return early, rotation schedule has changed
+		return staticCredsMeta, resp, nil
+	}
+
+	logger = logger.WithValues(
+		"staticCredsMeta", staticCredsMeta,
+		"lastSyncStaticCredsMeta", lastSyncStaticCredsMeta,
+		"ttl", staticCredsMeta.TTL,
+		"inLastSyncRotation", inLastSyncRotation,
+	)
+
+	bo := backoff.NewExponentialBackOff(
+		// typically the rotation period is 5s, so it should be safe to double that.
+		// Ideally we could use the rotation's TTL value here, but that value is not
+		// considered to be reliable to the TTL roll-over bug that might exist in the database
+		// secrets engine.
```

Comment:
Yeah that is a great question. We will be fixing the issue in Vault, just no ETA yet. I think it's okay to provide a best effort workaround for those users affected, especially since this workaround is to address a support escalation.